builtins/amp-img.js

```diff
@@ -178,6 +203,21 @@ export class AmpImg extends BaseElement {
     return promise;
   }
 
+  /**
+   * Sets the img src to the first url in the srcset if srcset is defined but
+   * src is not.
+   * @private
+   */
+  guaranteeSrcForSrcsetUnsupportedBrowsers_() {
+    if (!this.img_.hasAttribute('src') && 'srcset' in this.img_ == false) {
+      const srcset = this.element.getAttribute('srcset');
+      const srcseturl = srcsetRegex.exec(srcset);
```

Comment:
OOOOoooo, @aghassemi's comment reminded me of something else:

Non-exhaustive use of global regexes is dangerous! Global regexes have state (`lastIndex`) which is used to find the next match in a series. `#match` is exhaustive, so it's fine. But, `#exec` isn't, so it leaves a new `lastIndex` on the regex after matching, causing mayhem!

```js
srcsetRegex.lastIndex; // => 0
srcsetRegex.exec('test.jpg'); // => ["test.jpg", "test.jpg", undefined, undefined]

srcsetRegex.lastIndex; // => 8
srcsetRegex.exec('test2.jpg'); // => ["g", "g", undefined, undefined]
```

You have to manually reset the `lastIndex` yourself afterwards! Or better yet, export a helper method that'll do the `#exec` and reset it for you.

Reply:
wow, TIL & scary! (current ToT about making interfaces predictable perfectly applies here)

Reply:
oh wow.